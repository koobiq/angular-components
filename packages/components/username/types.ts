/**
 * Keys for formatting username parts in short (initial) or full form.
 */
export enum KbqUsernameFormatKey {
    /**
     * Short form of the first name (e.g., "John" → "J")
     */
    FirstNameShort = 'f',

    /**
     * Full form of the first name (e.g., "John")
     */
    FirstNameFull = 'F',

    /**
     * Short form of the middle name (e.g., "Henry" → "H")
     */
    MiddleNameShort = 'm',

    /**
     * Full form of the middle name (e.g., "Henry")
     */
    MiddleNameFull = 'M',

    /**
     * Short form of the last name (e.g., "Doe" → "D")
     */
    LastNameShort = 'l',

    /**
     * Full form of the last name (e.g., "Doe")
     */
    LastNameFull = 'L'
}

/**
 * Maps each format key to a property name in the user profile object.
 * Allows flexible formatting regardless of profile field names.
 */
export type KbqFormatKeyToProfileMapping<T = any> = {
    [key in KbqUsernameFormatKey]: keyof T;
};

export type KbqUsernameMode = 'stacked' | 'inline' | 'text';
