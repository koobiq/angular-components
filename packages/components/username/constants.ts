import { InjectionToken } from '@angular/core';
import { KbqFormatKeyToProfileMapping, KbqFormatKeyToProfileMappingExtended, KbqUsernameFormatKey } from './types';

/** Default name format: Last name full, first and middle as initials. */
export const kbqDefaultFullNameFormatCustom = 'L f. m.';
/** Default name format: Last name full, first and middle as initials. */
export const kbqDefaultFullNameFormat = 'lf.m.';

/**
 * Throws an error when no profile field mapping is provided to the username pipe.
 * @docs-private
 */
export function KbqMappingMissingError() {
    return new Error('KbqUsernamePipe: profile field mapping is required but was not provided.');
}

/**
 * Injection token for providing a global username format-to-profile mapping.
 */
export const KBQ_PROFILE_MAPPING = new InjectionToken<
    KbqFormatKeyToProfileMapping | KbqFormatKeyToProfileMappingExtended
>('KBQ_PROFILE_MAPPING', {
    factory: () =>
        ({
            [KbqUsernameFormatKey.FirstNameShort]: 'firstName',
            [KbqUsernameFormatKey.MiddleNameShort]: 'middleName',
            [KbqUsernameFormatKey.LastNameShort]: 'lastName',
            [KbqUsernameFormatKey.Dot]: undefined
        }) satisfies KbqFormatKeyToProfileMapping<{ firstName: string; middleName: string; lastName: string }>
});
