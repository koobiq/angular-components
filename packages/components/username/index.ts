export * from './constants';
export * from './module';
export * from './types';
export * from './username';
export * from './username.pipe';
