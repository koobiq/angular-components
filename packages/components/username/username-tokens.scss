:where(.kbq-username) {
    --kbq-username-primary-color: var(--kbq-foreground-contrast);
    --kbq-username-secondary-color: var(--kbq-foreground-contrast-secondary);
    --kbq-username-vertical-gap: var(--kbq-size-xxs);
    --kbq-username-horizontal-gap: var(--kbq-size-xxs);
}
