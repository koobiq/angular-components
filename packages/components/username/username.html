@let profile = userInfo();
@let fullName = profile | kbqUsername: fullNameFormat();

@if (!isCompact()) {
    @if (hasFullName()) {
        <span kbq-title class="kbq-username__name">{{ fullName }}</span>
    }

    @if (profile?.login) {
        <span kbq-title class="kbq-username__login">
            {{ profile?.login }}

            @if (profile?.site) {
                <span class="kbq-username__site">({{ profile?.site }})</span>
            }
        </span>
    }
} @else {
    <span kbq-title class="kbq-username__name">
        @if (hasFullName()) {
            <span>{{ fullName }}</span>
        }
        @if (!hasFullName() && profile?.login) {
            {{ profile?.login }}
        }

        @if (profile?.site) {
            <span class="kbq-username__site">&nbsp;({{ profile?.site }})</span>
        }
    </span>
}
