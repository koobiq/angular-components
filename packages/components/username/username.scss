@use '../core/styles/common';
@use '../core/styles/common/tokens';

// Targets the layout container: either the root or custom view if present.
@mixin _kbq-username-content-wrapper {
    &:not(:has(> .kbq-username__custom-view)),
    & > .kbq-username__custom-view {
        @content;
    }
}

@mixin _kbq-username-theme() {
    & {
        color: var(--kbq-username-primary-color);

        .kbq-username__secondary-hint,
        .kbq-username__primary + .kbq-username__secondary {
            color: var(--kbq-username-secondary-color);
        }
    }

    &.kbq-username_error {
        --kbq-username-primary-color: var(--kbq-foreground-error);
        --kbq-username-secondary-color: var(--kbq-foreground-error);
    }

    &.kbq-username_accented {
        --kbq-username-primary-color: var(--kbq-foreground-contrast);
        --kbq-username-secondary-color: var(--kbq-foreground-contrast);
    }
}

@mixin _kbq-username-typography() {
    & {
        @include tokens.kbq-typography-level-to-styles_css-variables(typography, text-normal);
    }

    &.kbq-username_accented {
        .kbq-username__primary {
            @include tokens.kbq-typography-level-to-styles_css-variables(typography, text-normal-strong);
        }
    }
}

.kbq-username {
    @include _kbq-username-theme();
    @include _kbq-username-typography();

    &.kbq-username_stacked {
        @include _kbq-username-content-wrapper {
            display: flex;
            flex-flow: column nowrap;
            justify-content: flex-start;
            align-items: normal;

            gap: var(--kbq-username-vertical-gap);
        }
    }

    &.kbq-username_inline:not(.kbq-username_inherit) {
        @include _kbq-username-content-wrapper {
            display: flex;
            flex-flow: row nowrap;
            justify-content: flex-start;
            align-items: baseline;

            .kbq-username__primary + .kbq-username__secondary {
                @include common.rtl-prop(margin-left, margin-right, var(--kbq-username-horizontal-gap), 0);
            }
        }
    }

    &.kbq-username_text {
        @include _kbq-username-content-wrapper {
            display: inline;
        }
    }

    &.kbq-username_stacked,
    &.kbq-username_inline {
        @include _kbq-username-content-wrapper {
            .kbq-username__primary,
            .kbq-username__secondary {
                @include common.kbq-truncate-line();
            }
        }
    }

    &.kbq-username_inherit {
        @include _kbq-username-content-wrapper {
            display: inline;
            color: inherit;

            .kbq-username__secondary,
            .kbq-username__secondary-hint {
                color: inherit;
            }
        }
    }
}

.kbq-link .kbq-username {
    color: inherit;

    .kbq-username__secondary,
    .kbq-username__secondary-hint {
        color: inherit;
    }
}
