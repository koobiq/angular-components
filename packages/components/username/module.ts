import { NgModule } from '@angular/core';
import { KbqUsername, KbqUsernameCustomView, KbqUsernamePrimary, KbqUsernameSecondary } from './username';
import { KbqUsernameCustomPipe, KbqUsernamePipe } from './username.pipe';

const COMPONENTS = [
    KbqUsername,
    KbqUsernameCustomView,
    KbqUsernamePrimary,
    KbqUsernameSecondary,
    KbqUsernameCustomPipe,
    KbqUsernamePipe
];

@NgModule({
    imports: COMPONENTS,
    exports: COMPONENTS
})
export class KbqUsernameModule {}
