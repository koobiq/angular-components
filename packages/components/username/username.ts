import {
    booleanAttribute,
    ChangeDetectionStrategy,
    Component,
    computed,
    contentChild,
    Directive,
    input,
    ViewEncapsulation
} from '@angular/core';
import { KbqTitleModule } from '@koobiq/components/title';
import { kbqDefaultFullNameFormat } from './constants';
import { KbqUsernameMode, KbqUsernameStyle } from './types';
import { KbqUsernamePipe } from './username.pipe';

const baseClass = 'kbq-username';

type Profile = {
    firstName?: string;
    lastName?: string;
    middleName?: string;
    login?: string;
    site?: string;
};

/** Styles the primary part of the username (e.g. full name). */
@Directive({
    selector: '[kbqUsernamePrimary]',
    exportAs: 'kbqUsernamePrimary',
    standalone: true,
    host: {
        class: `${baseClass}__primary`
    }
})
export class KbqUsernamePrimary {}

/** Styles the secondary part. */
@Directive({
    selector: '[kbqUsernameSecondary]',
    exportAs: 'kbqUsernameSecondary',
    standalone: true,
    host: {
        class: `${baseClass}__secondary`
    }
})
export class KbqUsernameSecondary {}

/** Styles a secondary hint. */
@Directive({
    selector: '[kbqUsernameSecondaryHint]',
    exportAs: 'kbqUsernameSecondaryHint',
    standalone: true,
    host: {
        class: `${baseClass}__secondary-hint`
    }
})
export class KbqUsernameSecondaryHint {}

/** Custom content for `KbqUsername`, overrides default view. */
@Directive({
    selector: 'kbq-username-custom-view, [kbq-username-custom-view]',
    standalone: true,
    exportAs: 'kbqUsernameCustomView',
    host: {
        class: `${baseClass}__custom-view`
    }
})
export class KbqUsernameCustomView {}

/**
 * Displays a user's name based on profile data.
 * Supports different display modes and visual styles.
 * A custom view can be provided via `<kbq-username-custom-view>` for full control over the output.
 * Accepts input profile data and optional formatting options.
 */
@Component({
    selector: 'kbq-username',
    standalone: true,
    exportAs: 'kbqUsername',
    imports: [
        KbqTitleModule,
        KbqUsernamePipe,
        KbqUsernamePrimary,
        KbqUsernameSecondary,
        KbqUsernameSecondaryHint
    ],
    templateUrl: './username.html',
    styleUrls: ['./username.scss', './username-tokens.scss'],
    host: {
        class: baseClass,
        '[class]': 'class()'
    },
    encapsulation: ViewEncapsulation.None,
    changeDetection: ChangeDetectionStrategy.OnPush
})
export class KbqUsername {
    /** User profile data used for display. */
    readonly userInfo = input<Profile>();
    /** Enables compact display mode */
    readonly isCompact = input(false, { transform: booleanAttribute });
    /** Format string for rendering the user's full name. */
    readonly fullNameFormat = input<string>(kbqDefaultFullNameFormat);
    /**
     * Display mode of the username.
     * @default inline
     */
    readonly mode = input<KbqUsernameMode>('inline');
    /**
     * Visual style of the username.
     * @default default
     */
    readonly type = input<KbqUsernameStyle>('default');
    /** Custom projected view for username rendering. */
    protected readonly customView = contentChild(KbqUsernameCustomView);

    /** @docs-private */
    protected readonly hasFullName = computed(() => {
        const profile = this.userInfo();

        if (!profile) return false;

        return profile?.lastName && profile?.firstName;
    });

    /** @docs-private */
    protected readonly class = computed(() => {
        return [this.type(), this.mode()].map((modificator) => `${baseClass}_${modificator}`).join(' ');
    });
}
