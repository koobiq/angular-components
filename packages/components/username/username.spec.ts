import { ChangeDetectionStrategy, Component, Type } from '@angular/core';
import { ComponentFixture, inject, TestBed } from '@angular/core/testing';
import { By } from '@angular/platform-browser';
import { KbqUsernameModule } from './module';
import {
    KbqFormatKeyToProfileMapping,
    KbqFormatKeyToProfileMappingExtended,
    KbqUsernameFormatKey,
    KbqUsernameMode,
    KbqUsernameStyle
} from './types';
import { KbqUsername, KbqUsernameCustomView } from './username';
import { KbqUsernameCustomPipe, KbqUsernamePipe } from './username.pipe';

const createComponent = <T>(component: Type<T>, providers: any[] = []): ComponentFixture<T> => {
    TestBed.configureTestingModule({ imports: [component], providers }).compileComponents();
    const fixture = TestBed.createComponent<T>(component);

    fixture.autoDetectChanges();

    return fixture;
};

type ExampleUser = {
    firstName?: string;
    lastName?: string;
    middleName?: string;
};

const mockProfile: ExampleUser = {
    firstName: 'Alice',
    middleName: 'Bishop',
    lastName: 'Carter'
};

describe(KbqUsernamePipe.name, () => {
    const mockMapping: KbqFormatKeyToProfileMapping<ExampleUser> = {
        [KbqUsernameFormatKey.FirstNameShort]: 'firstName',
        [KbqUsernameFormatKey.MiddleNameShort]: 'middleName',
        [KbqUsernameFormatKey.LastNameShort]: 'lastName',
        [KbqUsernameFormatKey.Dot]: undefined
    };

    let pipe: KbqUsernamePipe<any>;

    beforeEach(() => {
        TestBed.configureTestingModule({
            imports: [KbqUsernameModule],
            providers: [
                KbqUsernamePipe
            ]
        }).compileComponents();
    });

    beforeEach(inject([KbqUsernamePipe], (p: KbqUsernamePipe<any>) => {
        pipe = p;
    }));

    it('should format full name using default format', () => {
        const result = pipe.transform(mockProfile, undefined, mockMapping);

        expect(result).toBe('Carter A. B.');
    });

    it('should return empty string for empty profile', () => {
        const result = pipe.transform(null as any, 'FML', mockMapping);

        expect(result).toBe('');
    });

    it('should return empty string if profile is not an object', () => {
        const result = pipe.transform([], 'FML', mockMapping);

        expect(result).toBe('');
    });

    it('should skip irrelevant letters in format', () => {
        const irrelevantLetter = 's';
        const result = pipe.transform(mockProfile, `lf.m.${irrelevantLetter}`, mockMapping);

        expect(result.includes(irrelevantLetter)).toBeFalsy();
    });
});

describe(KbqUsernamePipe.name, () => {
    let pipe: KbqUsernameCustomPipe<any>;
    const mockMapping: KbqFormatKeyToProfileMappingExtended<ExampleUser> = {
        [KbqUsernameFormatKey.FirstNameShort]: 'firstName',
        [KbqUsernameFormatKey.FirstNameFull]: 'firstName',
        [KbqUsernameFormatKey.MiddleNameShort]: 'middleName',
        [KbqUsernameFormatKey.MiddleNameFull]: 'middleName',
        [KbqUsernameFormatKey.LastNameShort]: 'lastName',
        [KbqUsernameFormatKey.LastNameFull]: 'lastName',
        [KbqUsernameFormatKey.Dot]: undefined
    };

    beforeEach(() => {
        TestBed.configureTestingModule({
            imports: [KbqUsernameModule],
            providers: [
                KbqUsernameCustomPipe
            ]
        }).compileComponents();
    });

    beforeEach(inject([KbqUsernameCustomPipe], (p: KbqUsernameCustomPipe<any>) => {
        pipe = p;
    }));

    it('should format full name using default format', () => {
        const result = pipe.transform(mockProfile, undefined, mockMapping);

        expect(result).toBe('Carter A. B.');
    });

    it('should format full name using custom format', () => {
        const formats = ['F m l', 'F m. l.', 'Fm.l.', 'f m l', 'f.m.l.', 'F\u2009m.\u2009l.'];

        expect(formats.map((format) => pipe.transform(mockProfile, format, mockMapping))).toMatchSnapshot();
    });

    it('should return empty string for empty profile', () => {
        const result = pipe.transform(null as any, 'FML', mockMapping);

        expect(result).toBe('');
    });

    it('should return empty string if profile is not an object', () => {
        const result = pipe.transform([], 'FML', mockMapping);

        expect(result).toBe('');
    });

    it('should add irrelevant letters in format', () => {
        const irrelevantLetter = 's';
        const result = pipe.transform(mockProfile, `lf.m.${irrelevantLetter}`, mockMapping);

        expect(result.includes(irrelevantLetter)).toBeTruthy();
    });
});

describe(KbqUsername.name, () => {
    it('should use default input values', () => {
        const { debugElement } = createComponent(TestComponent);

        expect(debugElement.query(By.directive(KbqUsername)).classes).toMatchSnapshot();
    });

    it('should use custom view instead default if provided', () => {
        const fixture = createComponent(CustomView);

        fixture.detectChanges();

        expect(fixture.debugElement.query(By.css('.kbq-username__primary'))).toBeFalsy();
    });
});

@Component({
    selector: 'test-component',
    standalone: true,
    imports: [
        KbqUsername
    ],
    template: `
        <kbq-username
            [userInfo]="userInfo"
            [fullNameFormat]="fullNameFormat"
            [isCompact]="isCompact"
            [mode]="selectedMode"
            [type]="selectedType"
        />
    `,
    changeDetection: ChangeDetectionStrategy.OnPush
})
export class TestComponent {
    userInfo: any = {
        firstName: 'firstName',
        middleName: 'MiddleName',
        lastName: 'LastName',
        login: 'login'
    };
    selectedMode: KbqUsernameMode = 'inline';
    selectedType: KbqUsernameStyle = 'default';
    isCompact = false;
    fullNameFormat = 'f.m.l';
}

@Component({
    selector: 'test-component',
    standalone: true,
    imports: [
        KbqUsername,
        KbqUsernameCustomView
    ],
    template: `
        <kbq-username [userInfo]="userInfo" [isCompact]="isCompact" [mode]="selectedMode" [type]="selectedType">
            <kbq-username-custom-view>Test</kbq-username-custom-view>
        </kbq-username>
    `,
    changeDetection: ChangeDetectionStrategy.OnPush
})
export class CustomView {
    userInfo: any = {
        firstName: 'firstName',
        middleName: 'MiddleName',
        lastName: 'LastName',
        login: 'login'
    };
    selectedMode: KbqUsernameMode = 'inline';
    selectedType: KbqUsernameStyle = 'default';
    isCompact = false;
}
