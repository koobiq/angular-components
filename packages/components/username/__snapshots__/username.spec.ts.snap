// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`KbqUsername should use default input values 1`] = `
{
  "kbq-username": true,
  "kbq-username_default": true,
  "kbq-username_inline": true,
}
`;

exports[`KbqUsernamePipe should format full name using custom format 1`] = `
[
  "Alice B C",
  "Alice B. C.",
  "AliceB.C.",
  "A B C",
  "A.B.C.",
  "Alice B. C.",
]
`;
